libcontainer/process_linux.go

```diff
@@ -229,13 +231,58 @@ func (p *initProcess) start() (err error) {
 	if err := p.sendConfig(); err != nil {
 		return newSystemError(err)
 	}
-	// wait for the child process to fully complete and receive an error message
-	// if one was encoutered
-	var ierr *genericError
-	if err := json.NewDecoder(p.parentPipe).Decode(&ierr); err != nil && err != io.EOF {
+
+	var (
+		procSync syncType
+		sentRun  bool
+		ierr     *genericError
+	)
+
+loop:
+	for {
+		if err := json.NewDecoder(p.parentPipe).Decode(&procSync); err != nil {
+			if err == io.EOF {
+				break loop
+			}
+			return newSystemError(err)
+		}
+
+		switch procSync {
+		case procStart:
+			break loop
+		case procReady:
+			if err := p.manager.SetLate(); err != nil {
```

Comment:
I've tested this out. The only issue is with `KernelMemory`, which requires `Set()` to be run before the cgroup is joined (kernel denies setting kmemcg when its not empty, and we can't join later for security reasons). I'll just allow this (and only this) option to be set during `Apply()` and the rest will be applied later.